### PR TITLE
Migrate to vue3

### PR DESCRIPTION
**Pull request by SamChou19815**
### Summary 

While no one is coding, let's try a migration to vue3...

Most changes are quite boring: in most component, you just change `Vue.extend` to `defineComponent`.

Some interesting changes:
- Entry point changes according to: https://v3.vuejs.org/guide/introduction.html#getting-started
- VueX store changes according to: https://next.vuex.vuejs.org/guide/
- VueRouter changes according to: https://next.router.vuejs.org/guide/#router-link
- VueDraggable changes according to: https://github.com/SortableJS/vue.draggable.next#migrate-from-vue-2-version
- VueGTag changes according to: https://matteo-gabriele.gitbook.io/vue-gtag/v/next/
- `click-outside` changes according to: https://stackoverflow.com/questions/63869859/detect-click-outside-element-on-vue-3

### Test Plan 

Just migrated... Let's test if everything still works.

### CODE REVIEW

File: src/components/BottomBar/BottomBarCourseReview.vue
Comment (willespencer) [inline src/components/BottomBar/BottomBarCourseReview.vue:121]: Did vue 3 require adding these types or did you just notice this?

File: src/components/Modals/NewCourse/CourseSelector.vue
Comment (willespencer) [inline src/components/Modals/NewCourse/CourseSelector.vue:64]: Unknown?

File: src/components/Requirements/IncompleteSubReqCourse.vue
Comment (willespencer) [inline src/components/Requirements/IncompleteSubReqCourse.vue:23]: Is this a new Vue 3 thing? Does this replace v-for loops somehow?

File: src/main.ts
Comment (willespencer) [inline src/main.ts:20]: Do we not use bootstrap anywhere on CoursePlan? I had thought the landing page used some bootstrap classes, but could be wrong

**Review by willespencer — CHANGES_REQUESTED** [review]
Looks good Sam! Left some clarification comments (mostly for my own knowledge about Vue 3). Also tested the site thoroughly (especially click outside and the different routes) and here are some comments with things that don't seem to work.

Here are a few bugs I did find though
- I'm not sure the new router code is working correctly? The policy page link in the footer just links back to the login page right now, and I can't seem to reach the 404 page. Not sure if this is just an issue with the hosting or not.

- Courses are not showing up in the requirements sidebar to drag over? Just the complete ones seem to display
image

- Relatedly, the clicking the reset button doesn't bring up the reset modal or do anything

- The add course buttons seem to not take up a whole course slot on both desktop and mobile in compact and non-compact views
image
image

- The scrolling I implemented for the new semester and self check year dropdowns don't seem to work anymore

- I don't know why any of your changes would affect this, but I'm noticing a bug with CUReviews? The data does not seem to load in on the bottom bar when a course is clicked for the first time, but after collapsing and reopening it, the bars fill up. This does not happen on courseplan.io though.

image

File: src/components/BottomBar/BottomBarCourseReview.vue
Comment (SamChou19815) [reply src/components/BottomBar/BottomBarCourseReview.vue:121]: Vue 3 seems to be more strict on these

File: src/components/Modals/NewCourse/CourseSelector.vue
Comment (SamChou19815) [reply src/components/Modals/NewCourse/CourseSelector.vue:64]: Yes... not sure why there is this error if I don't cast to unknown first

File: src/components/Requirements/IncompleteSubReqCourse.vue
Comment (SamChou19815) [reply src/components/Requirements/IncompleteSubReqCourse.vue:23]: No I think it's only specific to the draggable library

File: src/main.ts
Comment (SamChou19815) [reply src/main.ts:20]: We only use bootstrap css, but not vue components

**Review by SamChou19815 — COMMENTED** [review]
Fix some issues. See the above commits. I have limited time this week so I would appreciate if someone else can also help.

**Comment by SamChou19815** [comment]
Bug list status

- [x] I'm not sure the new router code is working correctly? The policy page link in the footer just links back to the login page right now, and I can't seem to reach the 404 page. Not sure if this is just an issue with the hosting or not.
- [x] Courses are not showing up in the requirements sidebar to drag over? Just the complete ones seem to display
- [x] Relatedly, the clicking the reset button doesn't bring up the reset modal or do anything
- [x] The add course buttons seem to not take up a whole course slot on both desktop and mobile in compact and non-compact.
- [ ] The scrolling I implemented for the new semester and self check year dropdowns don't seem to work anymore
- [x] I don't know why any of your changes would affect this, but I'm noticing a bug with CUReviews? The data does not seem to load in on the bottom bar when a course is clicked for the first time, but after collapsing and reopening it, the bars fill up. This does not happen on courseplan.io though.

**Comment by SamChou19815** [comment]
Looking at this breaking changes list: https://v3.vuejs.org/guide/migration/introduction.html
I think everything should be addressed now.

File: src/components/Semester/Semester.vue
Comment (willespencer) [inline src/components/Semester/Semester.vue:64]: Do we need to do this to any other in-line styles in the codebase?

File: src/components/Modals/Onboarding/OnboardingBasicSingleDropdown.vue
Comment (willespencer) [inline src/components/Modals/Onboarding/OnboardingBasicSingleDropdown.vue:92]: Can you make this same change to the code for the onboarding graduation year dropdown? I forgot I made that autoscroll last-minute as well.

**Review by willespencer — CHANGES_REQUESTED** [review]
All the bugs I found were fixed, thank you Sam! Just left two more comments, one small question and one small fix. I also think before merging this, @tcho6319 and/or other devs should take a look and try to find bugs, I'll bump this on slack.

File: src/components/Semester/Semester.vue
Comment (SamChou19815) [reply src/components/Semester/Semester.vue:64]: No I think it just applies to this library.

File: src/components/Modals/Onboarding/OnboardingBasicSingleDropdown.vue
Comment (tcho6319) [reply src/components/Modals/Onboarding/OnboardingBasicSingleDropdown.vue:92]: Yes +1 to what @willespencer said! Currently, when clicking the dropdown for graduation year in Onboarding, the dropdown doesn't autoscroll to 2021 and I get a type error appear in the console. Note that this doesn't show up when I click on any other dropdown in Onboarding. The graduation year does seem to save though.

https://user-images.githubusercontent.com/54298311/114028676-be228c80-9846-11eb-9314-50da6b64e55d.mov

File: src/components/Requirements/IncompleteSubReqCourse.vue
Comment (tcho6319) [inline src/components/Requirements/IncompleteSubReqCourse.vue:18]: Would it be better to use `v-model` here?

Screen Shot 2021-04-08 at 9 21 48 AM

File: src/router/index.ts
Comment (tcho6319) [inline src/router/index.ts:31]: Just curious, did the previous path `'/*'` not work?

File: tsconfig.json
Comment (tcho6319) [inline tsconfig.json:13]: Why would we want to skip type checking .d.ts files?

**Review by tcho6319 — APPROVED** [review]
Thank you so much for this migration and addressing Will's comments, Sam! Everything LGTM except for a type error bug that @willespencer already picked up on above and some questions that I had.

File: src/components/Requirements/IncompleteSubReqCourse.vue
Comment (SamChou19815) [reply src/components/Requirements/IncompleteSubReqCourse.vue:18]: To ensure that it's readonly.

File: src/router/index.ts
Comment (SamChou19815) [reply src/router/index.ts:31]: Yeah the old setup doesn't work. For reference, it's from this doc: https://next.router.vuejs.org/guide/essentials/dynamic-matching.html#reacting-to-params-changes

File: tsconfig.json
Comment (SamChou19815) [reply tsconfig.json:13]: Mostly to speed up things since we assume provided libraries' typedef is accurate. It's also suggested by the vue3 project creator.

**Review by willespencer — APPROVED** [review]
LGTM!

**Comment by SamChou19815** [comment]
## Summary for people who are interested in the migration

### Vue3 related major changes explanation

- A New Global API: `createApp`
  - Impacted our entry point `main.ts`, and various places to use plugins
- `defineComponent(...)` replaces `Vue.extend(...)`
  - Impacted all components. This is needed for tooling and TS support
- `v-model` usage on components has been reworked, replacing `v-bind.sync`
  - `v-model` is now syntactic sugar for `:modelValue=...` and `@input:modelValue=...`
  - Impacted `vue-draggable` and `ResetConfirmationModal` that uses `v-model`
- `ref inside v-for no longer register an array of refs`
  - Impacted scroll to functionality in dropdowns that previously depends on an array of refs.

### `VueRouter` changes

- New setup functions
- New syntax to declare catch-all 404 route

### `VueX` changes

New setup functions and plugin registration

### `vue-draggable` changes

- Need to wrap draggable elements inside `<template>`
- Need an `itemKey` for every element to be draggable
- Impacted `Semester.vue`, `IncompleteSubReq.vue` and `Requirements.vue` that has draggable course cards.

**Pull request merged**
